docs/proposals/job.md

```diff
@@ -0,0 +1,145 @@
+# Job Controller
+
+## Abstract
+A proposal for implementing a new controller - Job controller - which will be responsible
+for managing pod(s) that require to run-once to a completion, in contrast to what
+ReplicationController currently offers.
+
+Several existing issues and PRs were already created regarding that particular subject:
+* Job Controller [#1624](https://github.com/GoogleCloudPlatform/kubernetes/issues/1624)
+* New Job resource [#7380](https://github.com/GoogleCloudPlatform/kubernetes/pull/7380)
+
+
+## Use Cases
+1. Be able to start of one or several pods tracked as a single entity.
+1. Be able to implement basic batch oriented tasks.
+1. Be able to get the job status.
+1. Be able to limit the execution time for a job.
+1. Be able to specify the number of instances performing a job.
+1. Be able to specify triggers on job’s success/failure.
+
+
+## Motivation
+Jobs are needed for executing multi-pod computation to completion; a good example
+here would be the ability to implement any type of batch oriented tasks or a MapReduce
+or Hadoop style workload.
+
+
+## Implementation
+Job controller is similar to replication controller in that they manage pods.
+This implies they will follow the same controller framework that replication
+controllers already defined.  The biggest difference between a `Job` and a
+`ReplicationController` object is the purpose; `ReplicationController`
+ensures that a specified number of Pods are running at any one time, whereas
+`Job` is responsible for keeping the desired number of Pods to a completion of
+a task. This will be represented by the `RestartPolicy` which is required to
+always take value of `RestartPolicyNever` or `RestartOnFailure`.
+
+
+The new `Job` object will have the following content:
+
+```go
+// Job represents the configuration of a single job.
+type Job struct {
+    TypeMeta
+    ObjectMeta
+
+    // Spec is a structure defining the expected behavior of a job.
+    Spec JobSpec
+
+    // Status is a structure describing current status of a job.
+    Status JobStatus
+}
+
+// JobList is a collection of jobs.
+type JobList struct {
+    TypeMeta
+    ListMeta
+
+    Items []Job
+}
+```
+
+`JobSpec` structure is defined to contain all the information how the actual job execution
+will look like.
+
+```go
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
```

Comment:
I don't think it can do it's thing without a way to preempt job pods when I need to scale up a replication controller. IIUC, the only admission control mechanism that could conceivably do this now is the ResourceQuota. Going with implementing this in admission control, assume I restrict batch workloads to a specific namespace. If I update the namespace's ResourceQuota, will pods be killed or will no new pods be admitted? Suppose pods are killed (which I don't know is the case). Which ones? Do I need a namespace per batch job? Suppose I'm at the limit of my ResourceQuota, which job's pods get admitted? The first request to go through as resource becomes available? Implementation wise this also doesn't work well with the current ControllerExpectation mechanism (think spinlock vs notify). There's a lot to figure out for something that could be solved with a MaxParallelism in the controller manager.